Add basic error injector mechanism.

PengZheng opened the pull request:
It improves the malloc/realloc mock introduced by #459 in the following ways:

- It can be used together with various sanitizers. 
- The caller's address can be used to specify which call to inject error into.

In my day-time job, I use it to inject error successfully into various libraries, e.g. libc, openssl, sqlite, libuv.

PengZheng committed:
Add basic error injector mechanism.

An injector for malloc family is implemented.

PengZheng committed:
Merge branch 'master' of https://gitbox.apache.org/repos/asf/celix into feature/error_injector

pnoltes commented:
Nice, this cleans up the test code. I will have a look into this and see if I can also understand the mechanism behind this.

PengZheng commented on misc/error_injector/error_injector.h:
```diff
@@ -0,0 +1,86 @@
+}
+
+
+#define CELIX_EI_IMPL0(name)                                                                             \
```
This is jest one error pattern, in which zero return value is treated as an error.
Macros for other patterns (indeed, there are plenty of them) will be added when needed.

PengZheng replied:
This is jest one error pattern, in which zero return value is treated as an error. `IMPL0` really means error on zero return value.
Macros for other patterns (indeed, there are plenty of them) will be added when needed.

PengZheng commented on misc/error_injector/error_injector.h:
```diff
@@ -0,0 +1,86 @@
+            /* we can not use CELIX_EI_GET_CALLER(addr, name ## _caller_level) */                        \
+            switch(name ## _caller_level) {                                                              \
+            case 0:                                                                                      \
+                CELIX_EI_GET_CALLER(addr, 0);                                                            \
```
IIRC, constant is required here.

PengZheng commented on misc/error_injector/error_injector.h:
```diff
@@ -0,0 +1,86 @@
+    Dl_info dlinfo;                             \
+    void *dlret;                                \
+    dlret = __builtin_return_address((level));  \
+    assert(dladdr(dlret, &dlinfo));             \
```
`-rdynamic` is used to make `dladdr` work.  This tech works for release and debug build. 
However, there is a limitation, if the caller is a static function (not visible outside the current translation unit), it will not work.

PengZheng reviewed (commented):
Several inline comments are added to facilitate understanding.

PengZheng committed:
Fix memory over-read and leak in PSA#465

It fixed several other bugs revealed by this fix:

* memory leak in zmq receiver
* endianness bugs
* uninitialised `message.header.convertEndianess` in zmq sender

It also serves as an examples of utilizing error injector to achieve high line/branch coverage.

PengZheng committed:
Merge branch 'master' of https://gitbox.apache.org/repos/asf/celix into feature/error_injector

PengZheng committed:
Fix new gcc warning after upgrading CI.

PengZheng committed:
Fix compilation error introduced by previous commit.

PengZheng committed:
Disable unsupported compiler option for macOS.

PengZheng committed:
Disable unsupported compiler option for macOS.

PengZheng committed:
Use more portable way to specify linker options.

PengZheng commented on misc/error_injector/malloc/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+target_include_directories(malloc_ei PUBLIC ${CMAKE_CURRENT_LIST_DIR})
+target_link_libraries(malloc_ei PUBLIC error_injector)
+# It plays nicely with address sanitizer this way.
+target_link_options(malloc_ei INTERFACE LINKER:--wrap,malloc LINKER:--wrap,realloc LINKER:--wrap,calloc)
```
```
ld: unknown option: --wrap
```

We need to figure out how to do it on macOS.

PengZheng reviewed (commented):
Though linker option `--wrap==symbol` is unsupported on macOS, `error_injector` is still usable.
I need to know more on macOS's linker.

PengZheng committed:
Separate error injection tests from common tests and turn them off on macOS.

PengZheng committed:
Fix compilation errors for build-brew and build-apt.

PengZheng committed:
Use macos-latest for CI.

PengZheng committed:
Revert "Use macos-latest for CI." to optimize conan build time.

This reverts commit 415c131652536e8640bcdf97e6bea2654038f152.

PengZheng reviewed (commented):
It seems that there is no macOS equivalence of `--wrap` linker option. 
Doing error injection with address sanitizer turned on can be very useful to find resource leaks (and other memory errors) in error handling path, which frequently happens.  Therefore, I'd rather do it perfectly on Linux, leaving the Mac OS untouched. WDYT? @pnoltes 

Another reasonable approach may be worth trying:

1. Directly override target function by redefining it in corresponding error injector static library.
2. Use `dlsym(RTLD_NEXT)`/`dlsym(specific library name)` to access the real target function implementation. I think the `RTLD_NEXT` solution is fragile, while getting the target library file name (not only for standard library but also for any third party libray) for the latter to work could be challenging for CMake.
3. Use `pthread_once` to protect multi-thread access.

Without a Mac at hand, it is inconvenient to do further research and tests.

pnoltes commented:
> 

IMO support for mac is needed. The main reason for error inject is to ensure that the error handle has coverage and its logic is tested. I think this can be done using linux. Of course we will miss some corner cases, but I do not see a big issue.

pnoltes commented on misc/error_injector/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+# specific language governing permissions and limitations
+# under the License.
+
+add_library(error_injector INTERFACE error_injector.h)
```
nitpick: IMO this library can be subdir of the libs directory and also has as name `Celix::error_injector`

pnoltes commented on misc/error_injector/celix_properties/celix_properties_ei.cpp:
```diff
@@ -0,0 +1,29 @@
+  under the License.
+ */
+
+#include <celix_properties_ei.h>
```
nitpick: `#include "celix_properties_ei.h"`

PengZheng replied:
Oops! I should have followed Celix's existing convention, and it will be corrected.

In John Lakos's Large Scale C++, it is suggested that angle brackets instead of double quotes should be used exclusively in include directives. I follow his methods in my day-time job. However, existing convention should always take priority.

pnoltes commented on misc/error_injector/error_injector.h:
```diff
@@ -0,0 +1,84 @@
+ */
+
+#ifndef CELIX_ERROR_INJECTOR_H
+#define CELIX_ERROR_INJECTOR_H
```
nitpick: please rename to celix_error_injector.h and move it to a include dir.

pnoltes commented on misc/error_injector/error_injector.h:
```diff
@@ -0,0 +1,84 @@
+#include <assert.h>
+#include <dlfcn.h>
+#include <stddef.h>
+
```
Could this macro get some documentation. Something like:

```
/**
 * @brief Get the caller library address based on the return address of level frame up t he call stack.
 */
#define CELIX_EI_GET_CALLER(addr, level)                                   \
do {                                                                       \
    Dl_info dlinfo;                                                        \
    dladdr(__builtin_return_address(level), &dlinfo);                     \
    (addr) = dlinfo.dli_saddr;                                             \
} while(0)
```

pnoltes replied:
Or maybe a small readme how to use this (declare, define, etc)

PengZheng replied:
I think both are needed.

pnoltes commented on misc/error_injector/celix_properties/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+# specific language governing permissions and limitations
+# under the License.
+
+add_library(celix_properties_ei STATIC celix_properties_ei.cpp)
```
nitpick: In Celix most resent C++ sources files use the `cc` extension

pnoltes commented on misc/error_injector/malloc/malloc_ei.cpp:
```diff
@@ -0,0 +1,43 @@
+#include <malloc_ei.h>
+
+extern "C" {
+void *__real_malloc(size_t);
```
Just to check. the symbol `__real_malloc` which pont to `malloc` due to ld `--wrap`?
And the any usage of `malloc` will result in symbol `__warp_malloc` due to ld `--wrap`?

PengZheng replied:
Yes, that's the way it works.

pnoltes commented on bundles/pubsub/pubsub_protocol/pubsub_protocol_lib/gtest/src/PS_WP_common_ei_tests.cc:
```diff
@@ -0,0 +1,233 @@
+
+    //Scenario: No mem with no pre-allocated data
+    //Given (mocked) realloc is forced to return NULL
+    celix_ei_expect_realloc((void *)pubsubProtocol_encodeMetadata, 0, nullptr);
```
Just to check. 

So the first call to realloc from pubsubProtocol_encodeMetadata (0 level deep) should fail and return a nullptr?

PengZheng replied:
Yes.

pnoltes commented on bundles/pubsub/pubsub_protocol/pubsub_protocol_lib/gtest/src/PS_WP_common_ei_tests.cc:
```diff
@@ -0,0 +1,233 @@
+    auto len = strlen(data);
+    pubsubProtocol_writeInt((unsigned char*)data, 0, message.header.convertEndianess, 3);
+    for (int i = 0; i < 6; ++i) {
+        celix_ei_expect_calloc((void *)pubsubProtocol_decodeMetadata, 0, nullptr, i+1);
```
Just to check.

So the `i+1` call to realloc from pubsubProtocol_encodeMetadata (0 level deep) should fail and return a nullptr?

PengZheng replied:
Correct. So that errors are injected in all call-sites.

pnoltes commented on bundles/pubsub/pubsub_protocol/pubsub_protocol_lib/gtest/CMakeLists.txt:
```diff
@@ -17,10 +17,16 @@
\ No newline at end of file
+setup_target_for_coverage(celix_pswp_common_tests SCAN_DIR ..)
+
+if (LINKER_WRAP_SUPPORTED)
```
Nice. Yeah so only if wrap is supported this is used and this includes the coverage build.

pnoltes reviewed (changes requested):
Very interesting PR. 
Quite some new stuff for me, but looks a very usable way of error injections in a fine grained manner.

I do have some remarks.

PengZheng committed:
Merge branch 'master' into feature/error_injector

PengZheng committed:
Project structure adjustments in compliance with existing convention.

PengZheng committed:
Add documentation for error injector.

pnoltes commented on libs/error_injector/README.md:
```diff
@@ -0,0 +1,81 @@
+limitations under the License.
+-->
+
+# Celix Error Injector
```
:+1: for updating the documentation.

Maybe a list can be added for which functions error injection is already prepared (`malloc` and `celix_properties_create`)?

PengZheng replied:
I'll add a table.

pnoltes commented on libs/error_injector/celix_properties/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# specific language governing permissions and limitations
+# under the License.
+
+add_library(properties_ei STATIC src/celix_properties_ei.cc)
```
Is there a reason why properties_ei and malloc_ei cannot be part of the error_injector library?

I understand that the current error_injector (interface) library provides the mechanisme to create error injection for functions, but I think we if we use error injection a single linkage setup is a bit user friendly.

pnoltes commented:
Overall LGTM, I added 2 small but not critical remarks

PengZheng committed:
Minor documentation improvement.
